From: Michael Tretter <m.tretter@pengutronix.de>
Date: Wed, 28 Jan 2026 22:25:23 +0100
Subject: [PATCH] tee: optee: add system_data driver

system_data is a OP-TEE TA that allows to store variables. Different
types of variables are supported:

- unrestricted
- write-once
- rollback-protected

write-once variables are allowed to be written exactly once.
rollback-protected variables have a (usually 8 byte) counter at the
start of the variable. Updates are only allowed when this counter is
increased with the write. Unrestricted variables do not have any
restrictions.

Upstream-Status: Pending

Signed-off-by: Michael Tretter <m.tretter@pengutronix.de>
Signed-off-by: Sascha Hauer <s.hauer@pengutronix.de>
---
 drivers/tee/optee/Kconfig          |   7 +
 drivers/tee/optee/Makefile         |   1 +
 drivers/tee/optee/system_data.c    | 273 +++++++++++++++++++++++++++++++++++++
 drivers/tee/optee/system_data_ta.h |  41 ++++++
 include/system_data.h              |  32 +++++
 5 files changed, 354 insertions(+)
 create mode 100644 drivers/tee/optee/system_data.c
 create mode 100644 drivers/tee/optee/system_data_ta.h
 create mode 100644 include/system_data.h

diff --git a/drivers/tee/optee/Kconfig b/drivers/tee/optee/Kconfig
index 73e77d0559c4..ad200b3a9e2a 100644
--- a/drivers/tee/optee/Kconfig
+++ b/drivers/tee/optee/Kconfig
@@ -47,6 +47,13 @@ config OPTEE_RKSECURE
 	  Enable this option to use the Rockchip Secure Boot PTA to read or
 	  write the secure boot OTP fuses of an rk3588 SoC.
 
+config OPTEE_SYSTEM_DATA
+	bool "Secure Data TA support"
+	depends on OPTEE
+	help
+	  Enable this option to use the System Data TA to store and read
+	  rollback protected data from the secure storage.
+
 endif
 
 config OF_FIXUP_OPTEE
diff --git a/drivers/tee/optee/Makefile b/drivers/tee/optee/Makefile
index e89edc876d6f..8d9caadef350 100644
--- a/drivers/tee/optee/Makefile
+++ b/drivers/tee/optee/Makefile
@@ -9,3 +9,4 @@ optee-objs += smc_abi.o
 obj-$(CONFIG_MCI_MMC_RPMB) += rpmb.o
 obj-$(CONFIG_OPTEE_AVB_PERSISTENT_VALUES) += avb.o
 obj-$(CONFIG_OPTEE_RKSECURE) += rksecure.o
+obj-$(CONFIG_OPTEE_SYSTEM_DATA) += system_data.o
diff --git a/drivers/tee/optee/system_data.c b/drivers/tee/optee/system_data.c
new file mode 100644
index 000000000000..8e28ac40a1b8
--- /dev/null
+++ b/drivers/tee/optee/system_data.c
@@ -0,0 +1,273 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) Pengutronix, Michael Tretter <m.tretter@pengutronix.de>
+ */
+
+#include <common.h>
+#include <init.h>
+#include <linux/minmax.h>
+#include <tee/optee.h>
+#include <linux/tee_drv.h>
+#include <linux/uuid.h>
+#include <uapi/linux/tee.h>
+
+#include <system_data.h>
+
+#include "optee_private.h"
+#include "system_data_ta.h"
+
+static const struct {
+	const char *name;
+	enum system_data_mode mode;
+} mode_names[] = {
+	{ "invalid", SYSTEM_DATA_MODE_INVALID },
+	{ "unrestricted", SYSTEM_DATA_MODE_UNRESTRICTED },
+	{ "rollback_protect", SYSTEM_DATA_MODE_ROLLBACK_PROTECT },
+	{ "write_once", SYSTEM_DATA_MODE_WRITE_ONCE }
+};
+
+const char *system_data_mode_name(enum system_data_mode mode)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(mode_names); i++)
+		if (mode == mode_names[i].mode)
+			return mode_names[i].name;
+
+	return ERR_PTR(-EINVAL);
+}
+
+const enum system_data_mode system_data_mode_by_name(const char *name)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(mode_names); i++)
+		if (strcmp(mode_names[i].name, name) == 0)
+			return mode_names[i].mode;
+
+	return SYSTEM_DATA_MODE_INVALID;
+}
+
+static int mode_to_ta(enum system_data_mode mode)
+{
+	/* Currently 1-to-1 mapping */
+	return mode;
+}
+
+struct system_data {
+	struct device *dev;
+	struct tee_context *ctx;
+	u32 session_id;
+};
+
+static struct system_data pvt_data;
+
+ssize_t system_data_read(struct system_data *system_data, void *buf,
+			 size_t count, const char *id)
+{
+	struct device *dev = system_data->dev;
+	struct tee_ioctl_invoke_arg inv_arg;
+	struct tee_param param[2];
+	struct tee_shm *id_shm;
+	size_t id_size;
+	struct tee_shm *data_shm;
+	size_t data_size;
+	struct system_data_object *data;
+	int res = 0;
+
+	dev_dbg(dev, "read %zu bytes from %s\n", count, id);
+
+	id_size = strlen(id);
+	id_shm = tee_shm_register_user_buf(system_data->ctx,
+					    (unsigned long)id, id_size);
+	if (IS_ERR(id_shm))
+		return -ENOMEM;
+
+	/* At least header + data, but wouldn't hurt to allocate a larger buffer */
+	data_size = sizeof(*data) + count;
+	data_shm = tee_shm_alloc_kernel_buf(system_data->ctx, data_size);
+	if (IS_ERR(data_shm)) {
+		res = -ENOMEM;
+		goto out;
+	}
+	data = data_shm->kaddr;
+
+	memset(param, 0, sizeof(param));
+	param[0].attr = TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_INPUT;
+	param[0].u.memref.shm = id_shm;
+	param[0].u.memref.size = strlen(id);
+
+	param[1].attr = TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_OUTPUT;
+	param[1].u.memref.shm = data_shm;
+	param[1].u.memref.size = data_size;
+
+	inv_arg.func = TA_SYSTEM_DATA_CMD_READ;
+	inv_arg.session = system_data->session_id;
+	inv_arg.num_params = ARRAY_SIZE(param);
+
+	res = tee_client_invoke_func(system_data->ctx, &inv_arg, param);
+	if (res)
+		goto out;
+
+	if (inv_arg.ret == TEEC_ERROR_SHORT_BUFFER) {
+		count = param[1].u.memref.size - sizeof(*data);
+		goto out;
+	} else if (inv_arg.ret) {
+		dev_err(dev, "invoke func failed with 0x%08x\n", inv_arg.ret);
+		res = -EIO;
+		goto out;
+	}
+
+	count = min(param[1].u.memref.size - sizeof(*data), count);
+	if (buf)
+		memcpy(buf, data->data, count);
+
+out:
+	tee_shm_free(data_shm);
+
+	if (res)
+		return res;
+	return count;
+}
+
+int system_data_write(struct system_data *system_data, const void *buf,
+		      size_t count, const char *id, enum system_data_mode mode)
+{
+	struct device *dev = system_data->dev;
+	struct tee_ioctl_invoke_arg inv_arg;
+	struct tee_param param[2];
+	struct tee_shm *id_shm;
+	size_t id_size;
+	struct tee_shm *data_shm;
+	size_t data_size;
+	struct system_data_object *data;
+	int res = 0;
+
+	dev_dbg(dev, "write %zu bytes to %s (mode %s)\n", count, id,
+		system_data_mode_name(mode));
+
+	id_size = strlen(id);
+	id_shm = tee_shm_register_user_buf(system_data->ctx,
+					    (unsigned long)id, id_size);
+	if (IS_ERR(id_shm))
+		return -ENOMEM;
+
+	/* At least header + data, but wouldn't hurt to allocate a larger buffer */
+	data_size = sizeof(*data) + count;
+	data_shm = tee_shm_alloc_kernel_buf(system_data->ctx, data_size);
+	if (IS_ERR(data_shm)) {
+		res = -ENOMEM;
+		goto out;
+	}
+	data = data_shm->kaddr;
+
+	memcpy(data->data, buf, count);
+	data->mode = mode_to_ta(mode);
+
+	memset(param, 0, sizeof(param));
+	param[0].attr = TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_INPUT;
+	param[0].u.memref.shm = id_shm;
+	param[0].u.memref.size = id_size;
+
+	param[1].attr = TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_INPUT;
+	param[1].u.memref.shm = data_shm;
+	param[1].u.memref.size = data_size;
+
+	inv_arg.func = TA_SYSTEM_DATA_CMD_WRITE;
+	inv_arg.session = system_data->session_id;
+	inv_arg.num_params = ARRAY_SIZE(param);
+
+	res = tee_client_invoke_func(system_data->ctx, &inv_arg, param);
+	if (res)
+		goto out;
+	if (inv_arg.ret) {
+		dev_err(dev, "invoke func failed with 0x%08x\n", inv_arg.ret);
+		res = -EIO;
+	}
+
+out:
+	tee_shm_free(data_shm);
+
+	return res;
+}
+
+struct system_data *system_data_get(void)
+{
+	if (pvt_data.ctx)
+		return &pvt_data;
+
+	return NULL;
+}
+
+static int optee_ctx_match(struct tee_ioctl_version_data *ver, const void *data)
+{
+	return ver->impl_id == TEE_IMPL_ID_OPTEE ? 1 : 0;
+}
+
+static int optee_system_data_probe(struct device *dev)
+{
+	struct tee_client_device *system_data_device = to_tee_client_device(dev);
+	struct tee_ioctl_open_session_arg sess_arg;
+	int ret, err = -ENODEV;
+
+	memset(&sess_arg, 0, sizeof(sess_arg));
+
+	/* Open context with TEE driver */
+	pvt_data.ctx = tee_client_open_context(NULL, optee_ctx_match, NULL,
+					       NULL);
+	if (IS_ERR(pvt_data.ctx))
+		return -ENODEV;
+
+	/* Open session with system-data Trusted App */
+	export_uuid(sess_arg.uuid, &system_data_device->id.uuid);
+	sess_arg.clnt_login = TEE_IOCTL_LOGIN_PUBLIC;
+	sess_arg.num_params = 0;
+
+	ret = tee_client_open_session(pvt_data.ctx, &sess_arg, NULL);
+	if ((ret < 0) || (sess_arg.ret != 0)) {
+		dev_err(dev, "tee_client_open_session failed, err: %x\n",
+			sess_arg.ret);
+		err = -EINVAL;
+		goto out_ctx;
+	}
+	pvt_data.session_id = sess_arg.session;
+
+	pvt_data.dev = dev;
+
+	return 0;
+
+out_ctx:
+	tee_client_close_context(pvt_data.ctx);
+	pvt_data.ctx = NULL;
+
+	return err;
+}
+
+static void optee_system_data_remove(struct device *dev)
+{
+	tee_client_close_session(pvt_data.ctx, pvt_data.session_id);
+	tee_client_close_context(pvt_data.ctx);
+	pvt_data.ctx = NULL;
+}
+
+static const struct tee_client_device_id optee_system_data_id_table[] = {
+	{UUID_INIT(0x41beaf7a, 0x2665, 0x4011,
+		   0xbe, 0xf6, 0x93, 0xff, 0x66, 0x45, 0x3c, 0x1b)},
+	{}
+};
+
+static struct tee_client_driver optee_system_data_driver = {
+	.id_table = optee_system_data_id_table,
+	.driver = {
+		.name = "optee-system-data",
+		.bus = &tee_bus_type,
+		.probe = optee_system_data_probe,
+		.remove = optee_system_data_remove,
+	},
+};
+
+static int optee_system_data_mod_init(void)
+{
+	return driver_register(&optee_system_data_driver.driver);
+}
+device_initcall(optee_system_data_mod_init);
diff --git a/drivers/tee/optee/system_data_ta.h b/drivers/tee/optee/system_data_ta.h
new file mode 100644
index 000000000000..e901492c8111
--- /dev/null
+++ b/drivers/tee/optee/system_data_ta.h
@@ -0,0 +1,41 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+#ifndef TA_SYSTEM_DATA_H
+#define TA_SYSTEM_DATA_H
+
+#include <types.h>
+
+#define TA_SYSTEM_DATA_UUID \
+	UUID_INIT(0x41beaf7a, 0x2665, 0x4011, \
+		  0xbe, 0xf6, 0x93, 0xff, 0x66, 0x45, 0x3c, 0x1b)
+
+/*
+ * TA_SYSTEM_DATA_CMD_READ - Read a stored object
+ *
+ * [in]     memref[0]    ID used the identify the persistent object
+ * [out]    memref[1]    system_data_object dumped from the persistent object
+ */
+#define TA_SYSTEM_DATA_CMD_READ			0
+
+/*
+ * TA_SYSTEM_DATA_CMD_WRITE - Persistently store an object, create if it doesn't exist
+ *
+ * [in]     memref[0]    ID used the identify the persistent object
+ * [in]     memref[1]    system_data_object to be written in the persistent object
+ */
+#define TA_SYSTEM_DATA_CMD_WRITE		1
+
+/* Types of write limitation */
+
+/* Unrestricted write access: */
+#define TA_SYSTEM_DATA_MODE_UNRESTRICTED		1
+/* Only allow writes that strictly increase the first differing byte: */
+#define TA_SYSTEM_DATA_MODE_ROLLBACK_PROTECT	2
+/* Immutable after initial creation: */
+#define TA_SYSTEM_DATA_MODE_WRITE_ONCE		3
+
+struct system_data_object {
+	uint32_t mode;
+	char data[];
+};
+
+#endif /*TA_SYSTEM_DATA_H*/
diff --git a/include/system_data.h b/include/system_data.h
new file mode 100644
index 000000000000..d2206783a173
--- /dev/null
+++ b/include/system_data.h
@@ -0,0 +1,32 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * Copyright (C) 2026 Pengutronix, Michael Tretter <m.tretter@pengutronix.de>
+ */
+
+#ifndef __SYSTEM_DATA_H
+#define __SYSTEM_DATA_H
+
+#include <linux/types.h>
+#include <driver.h>
+
+struct system_data;
+
+enum system_data_mode {
+	SYSTEM_DATA_MODE_INVALID = 0,
+	SYSTEM_DATA_MODE_UNRESTRICTED = 1,
+	SYSTEM_DATA_MODE_ROLLBACK_PROTECT = 2,
+	SYSTEM_DATA_MODE_WRITE_ONCE = 3,
+};
+
+const char *system_data_mode_name(enum system_data_mode mode);
+const enum system_data_mode system_data_mode_by_name(const char *name);
+
+int system_data_write(struct system_data *system_data,
+		      const void *buf, size_t count, const char *id,
+		      enum system_data_mode mode);
+ssize_t system_data_read(struct system_data *system_data,
+			 void *buf, size_t count, const char *id);
+
+struct system_data *system_data_get(void);
+
+#endif /* __SYSTEM_DATA_H */
